app/dashboard/src/providers/FeatureFlagsProvider.tsx

```diff
@@ -0,0 +1,118 @@
+/**
+ * @file
+ *
+ * Feature flags provider.
+ * Feature flags are used to enable or disable certain features in the application.
+ */
+import { useMount } from '#/hooks/mountHooks'
+import { useLocalStorage } from '#/providers/LocalStorageProvider'
+import LocalStorage from '#/utilities/LocalStorage'
+import type { ReactNode } from 'react'
+import { useEffect } from 'react'
+import { z } from 'zod'
+import { createStore, useStore } from 'zustand'
+
+declare module '#/utilities/LocalStorage' {
+  /**
+   * Local storage data structure.
+   */
+  interface LocalStorageData {
+    readonly featureFlags: z.infer<typeof FEATURE_FLAGS_SCHEMA>
+  }
+}
+
+export const FEATURE_FLAGS_SCHEMA = z.object({
+  enableMultitabs: z.boolean(),
+  enableAssetsTableBackgroundRefresh: z.boolean(),
+  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+  assetsTableBackgroundRefreshInterval: z.number().min(100),
+})
+
+LocalStorage.registerKey('featureFlags', { schema: FEATURE_FLAGS_SCHEMA })
+
+/**
+ * Feature flags store.
+ */
+export interface FeatureFlags {
+  readonly featureFlags: {
+    readonly enableMultitabs: boolean
+    readonly enableAssetsTableBackgroundRefresh: boolean
+    readonly assetsTableBackgroundRefreshInterval: number
+  }
+  readonly setFeatureFlags: <
+    Key extends keyof FeatureFlags['featureFlags'],
+    Value extends FeatureFlags['featureFlags'][Key],
+  >(
+    key: Key,
+    value: Value,
+  ) => void
+}
+
+const flagsStore = createStore<FeatureFlags>((set) => ({
+  featureFlags: {
+    enableMultitabs: false,
+    enableAssetsTableBackgroundRefresh: false,
+    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    assetsTableBackgroundRefreshInterval: 3_000,
+  },
+  setFeatureFlags: (key, value) => {
+    set(({ featureFlags }) => ({ featureFlags: { ...featureFlags, [key]: value } }))
+  },
+}))
+
+/**
+ * Hook to get all feature flags.
+ */
+export function useFeatureFlags() {
+  return useStore(flagsStore, (state) => state.featureFlags)
+}
+
+/**
+ * Hook to get a specific feature flag.
+ */
+export function useFeatureFlag<Key extends keyof FeatureFlags['featureFlags']>(
+  key: Key,
+): FeatureFlags['featureFlags'][Key] {
+  return useStore(flagsStore, ({ featureFlags }) => featureFlags[key])
+}
+
+/**
+ * Hook to set feature flags.
+ */
+export function useSetFeatureFlags() {
+  return useStore(flagsStore, ({ setFeatureFlags }) => setFeatureFlags)
+}
+
+/**
+ * Feature flags provider.
+ * Gets feature flags from local storage and sets them in the store.
+ * Also saves feature flags to local storage when they change.
+ */
+export function FeatureFlagsProvider({ children }: { children: ReactNode }) {
+  const { localStorage } = useLocalStorage()
+  const setFeatureFlags = useSetFeatureFlags()
+
+  useMount(() => {
+    const storedFeatureFlags = localStorage.get('featureFlags')
+
+    if (storedFeatureFlags) {
+      for (const [key, value] of Object.entries(storedFeatureFlags)) {
```

Comment:
consider using `unsafeEntries` instead

Reply:
I see that the only thing `unsafeEntries` does is add proper types, right?

Can we override typings for `ObjectConstructor` instead?

Reply:
i think that's an option, but i think it's preferred in general to have it separate as it is *technically* unsafe when there are extra keys not present in the type. it's not *always* the case but the idea is that it's common enough to pass e.g. an event or an asset, when the function accepts only a subset, that the cost of (auto-)importing the helper function is a small price to pay

or in other words: the "unsafe" being there is intentional, it's *supposed* to draw attention to it to make sure you've actually thought it through and have confirmed that it is safe (that the object is exactly the same shape as the type with no extra keys)

Reply:
worth noting that neither ts-reset nor better-typescript-lib seem to define an overload for this

Reply:
> worth noting that neither ts-reset nor better-typescript-lib seem to define an overload for this

Yeah, I checked both for that before leaving a comment here

>  the "unsafe" being there is intentional, it's supposed to draw attention to it to make sure you've actually thought it through and have confirmed that it is safe (that the object is exactly the same shape as the type with no extra keys)

Typescript does check that, doesn't it?
Here's a playground, and TS shows an error: https://www.typescriptlang.org/play/?ts=5.5.4#code/JYOwLgpgTgZghgYwgAgILIN4Chm+XALmRAFcBbAI2hzwqNMur2QXvKqiwF8stRJYiFACFkEAB6QQAEwDOaTDVzS2jTswiqO3XggD2IWWHxF0AXkXNCABgA0S5HTsPWyazqz7DxuslEXsKyJnZl8QvFdw5WD7DWCPAHoAKiTkACUIMBIoEGQwAAsUCHAoYAh5PRh8XL0KACsIBDAAOmQAVQA5AGVUADEAUWQDABsAT2QAd0Lc4GNgeQAHPVlZYAphlBg9KGqh+sbjMD1kfLgANwgcVIkwKDhkAGsIUdlWpISsCSWoYxgSECawAMyH+sjgMAg-RKZVkAB4ACpiSTFOR7BpNAB8AAoaLV0WAiPD7ABKIhQCBwaQjcYYZAAbQA0shQI9npVkPCALpkilUkBjekM2wcxmcznILh0p6jdlcuniwLIBIJZAAATAsgAtF8DtqoFBthz8vNmfJ+MUwECQHBhgL7hM7gsFtBkH8AZbgRNZvl8MhpMAYBDyeA8qNnc0aOSsjlkAB5fZNZoW0rlLF4g7EjxbHZYrxGekPYVncXs0HgyHQ1MUYnEyx4OBS8UWM7cIA

Reply:
I used the `unsafeObject` here, but I assume that with an override the result will be the same

Reply:
![CleanShot 2024-08-14 at 15 21 22@2x](https://github.com/user-attachments/assets/08ccc594-4614-47da-a1b2-f36d27086d4f)

Oh, I see

Reply:
@somebody1234 don't you think that narrowing the return type in `unsafeEntries` won't protect us from _mistakenly_ assigned excess properties? 

I think that using `Object.entries` is safer.

Reply:
> won't protect us from mistakenly assigned excess properties?

yeah, that's part of the reason why it's called unsafe - you'd want to use it *only* when you are 100% sure the type is correct - which does happen in *some* places in the codebase - e.g. when it's near the object literal where it is defined, or it is validated prior with e.g. zod

> I think that using Object.entries is safer.

is this true even with the (unsafe) type assertion here?